bin/mysql.sh

```diff
@@ -0,0 +1,386 @@
+        createDefaultUser  # create the default DB user
+        createReplUser     # create the replication user
+
+        # run user-defined files added to /etc/initdb.d in a child Docker image
+        for f in /etc/initdb.d/*; do
+            case "$f" in
+                *.sh)  echo "$0: running $f"; . "$f" ;;
+                *.sql) echo "$0: running $f"; "${mysql[@]}" < "$f" && echo ;;
+                *)     echo "$0: ignoring $f" ;;
+            esac
+        done
+
+        # remove the one-time root password (default behavior)
+        if [ ! -z "$MYSQL_ONETIME_PASSWORD" ]; then
+            "${mysql[@]}" <<-EOSQL
+                ALTER USER 'root'@'%' PASSWORD EXPIRE;
+EOSQL
+        fi
+
+        echo
+        echo 'MySQL init process done. Ready for start up.'
+        echo
+    fi
+}
+
+
+# ---------------------------------------------------------
+# Replication
+
+# Query Consul for the PRIMARY_NODE key
+getPrimaryNode() {
+    PRIMARY_NODE=$(curl -s http://${CONSUL}:8500/v1/kv/mysql-primary | \
+                          jq -r '.[].Value' | base64 --decode)
+}
+
+# Query Consul for healthy mysql nodes and check their ServiceID vs
+# PRIMARY_NODE. Sets PRIMARY_HOST to the IP address of a matching node.
+getPrimaryHost() {
+
+    if [[ -z ${PRIMARY_NODE} ]]; then
+        getPrimaryNode
+    fi
+
+    echo "Checking if primary (${PRIMARY_NODE}) is healthy..."
+
+    # providing a 30 second timeout here to avoid bootstrap failures on
+    # transient network paritions
+    for i in {30..0}; do
+        PRIMARY_HOST=$(curl -Ls --fail http://${CONSUL}:8500/v1/catalog/service/mysql?passing \
+            | jq -r 'map(select(.ServiceID == "'${PRIMARY_NODE}'")) | .[0].ServiceAddress')
+        if [[ ${PRIMARY_HOST} != "null" ]] && [[ -n ${PRIMARY_HOST} ]]; then
+            echo 'Found primary at:' ${PRIMARY_HOST}
+            break
+        fi
+        echo -ne '.'
+        sleep 1
+    done
+}
+
+# returns "true" on success or "false" on failure due to CAS
+markSelfPrimary() {
+    PRIMARY_NODE=mysql-$(hostname)
+    echo "Marking ${PRIMARY_NODE} as primary..."
+    for i in {30..0}; do
+        local ok=$(curl -XPUT -s --fail -d ${PRIMARY_NODE} \
+                        "http://${CONSUL}:8500/v1/kv/mysql-primary?cas=0" | jq -r '.')
+        if [[ ${ok} = "true" ]]; then
+            echo
+            break
+        fi
+        if [[ ${ok} = "false" ]]; then
+            # The only way this scenario appears is if we raced with
+            # another node. So we're going to just start over.
+            echo 'Tried to set this instance as primary but primary exists.'
+            echo 'Restarting bootstrap process.'
+            onStart
+        fi
+        echo -ne '.'
+        sleep 1
+    done
+}
+
+# Set up GITD-based replication to the primary; once this is set the
+# replica will automatically try to catch up with the primary by pulling
+# its entire binlog. This is comparatively slow but suitable for small
```

Comment:
At some point in a DB's age the bin logs have to be rotated out...this will only work for small DBs with little age.